lib/gateway.js

```diff
@@ -0,0 +1,250 @@
+    })
+    end(req, opts.body, done)
+  }
+
+  return {
+    request,
+    close
+  }
+}
+
+function createTypeMap (schemaDefinition) {
+  const parsedSchema = parse(schemaDefinition)
+  const typeMap = {}
+
+  for (const definition of parsedSchema.definitions) {
+    if (isTypeDefinitionNode(definition)) {
+      typeMap[definition.name.value] = new Set(definition.fields.map(field => field.name.value))
+    }
+  }
+
+  return typeMap
+}
+
+async function buildServiceMap (services) {
+  const serviceMap = {}
+
+  for (const service of services) {
+    const { name, ...opts } = service
+    const { request, close } = buildRequest(opts)
+    const url = new URL(opts.url)
+
+    const serviceConfig = {
+      request: function (opts) {
+        return new Promise((resolve, reject) => {
+          request({
+            url,
+            method: opts.method || 'POST',
+            body: opts.body,
+            headers: {
+              'content-type': 'application/json',
+              ...opts.headers
+            }
+          }, (err, response) => {
+            if (err) {
+              return reject(err)
+            }
+
+            if (response.statusCode === 200) {
+              response.stream.on('data', data => resolve({
+                statusCode: response.statusCode,
+                json: JSON.parse(data.toString())
+              }))
+            }
+          })
+        })
+      },
+      close,
+      init: async () => {
+        const response = await serviceConfig.request({
+          method: 'POST',
+          body: JSON.stringify({
+            query: `
+            query ServiceInfo {
+              _service {
+                sdl
+              }
+            }
+            `
+          })
+        })
+
+        const schemaDefinition = response.json.data._service.sdl
+
+        const typeMap = createTypeMap(schemaDefinition)
+
+        return {
+          schemaDefinition,
+          typeMap
+        }
+      }
+    }
+
+    serviceMap[service.name] = serviceConfig
+  }
+
+  await Promise.all(services.map(service => {
+    return serviceMap[service.name].init().then(({ schemaDefinition, typeMap }) => {
+      serviceMap[service.name].schemaDefinition = schemaDefinition
+      serviceMap[service.name].typeMap = typeMap
+      serviceMap[service.name].types = new Set(Object.keys(typeMap))
+    })
+  }))
+
+  return serviceMap
+}
+
+function isDefaultType (type) {
+  return [
+    '__Schema',
+    '__Type',
+    '__Field',
+    '__InputValue',
+    '__EnumValue',
+    '__Directive'
+  ].includes(type)
+}
+
+function makeResolveQuery (service) {
+  return function (root, args, context, info) {
+    const { fieldName, returnType, operation } = info
+
+    // Must deep-copy the operation because if we modify it (remove an item from the selection set)
+    // then the resolvers for those fields won't be executed
+    const currentOperation = JSON.parse(JSON.stringify(operation))
```

Comment:
This is terribly slow. Please check if you can use http://npm.im/rdfc

Reply:
I tried and it is not possible. The fields array is not a simple object and it throws a max call stack size exceeded error.

Reply:
It was possible to remove this in the lastest change